sgkit/stats/popgen.py

```diff
@@ -910,3 +914,156 @@ def Garud_H(
     )
 
     return conditional_merge_datasets(ds, new_ds, merge)
+
+
+@guvectorize(  # type: ignore
+    [
+        "void(float64[:], int32[:], uint8[:], float64[:])",
+        "void(float64[:], int64[:], uint8[:], float64[:])",
+    ],
+    "(n),(n),(c)->(c)",
+    nopython=True,
+    cache=True,
+)
+def _cohort_observed_heterozygosity(
+    hi: ArrayLike, cohorts: ArrayLike, _: ArrayLike, out: ArrayLike
+) -> None:  # pragma: no cover
+    """Generalized U-function for computing cohort observed heterozygosity.
+
+    Parameters
+    ----------
+    hi
+        Individual sample heterozygosity of shape (samples,).
+    cohorts
+        Cohort indexes for samples of shape (samples,).
+    _
+        Dummy variable of type `uint8` and shape (cohorts,) used to
+        define the number of cohorts.
+    out
+        Observed heterozygosity with shape (cohorts,) and values corresponding
+        to the mean individual heterozygosity of each cohort.
+    """
+    out[:] = 0
+    _[:] = 0
+    n_samples = len(hi)
+    n_cohorts = len(out)
+    for i in range(n_samples):
+        h = hi[i]
+        if not np.isnan(h):
+            c = cohorts[i]
+            out[c] += h
+            _[c] += 1
+    for j in range(n_cohorts):
+        n = _[j]
+        if n:
+            out[j] /= n
+        else:
+            out[j] = np.nan
+
+
+def observed_heterozygosity(
+    ds: Dataset,
+    *,
+    call_heterozygosity: Hashable = variables.call_heterozygosity,
+    merge: bool = True,
+) -> Dataset:
+    """Compute per cohort observed heterozygosity.
```

Comment:
Could either link to the documentation for the other heterozygosity function, or give the definition here again.